### PR TITLE
Differential privacy

### PR DESCRIPTION
**Pull request by caiom**
Implements differential privacy.

### CODE REVIEW

**Review by gugarosa — APPROVED** [review]

File: archai/nlp/datasets/corpus.py [hunk 1]
```diff
@@ -198,11 +200,15 @@ def _train_vocab(self)->None:
             self.vocab.load()
             logging.info(f'Vocab cache found and loaded for type {self.vocab_type} and size {self.vocab_size} from {self._vocab_cache_dir}.')
 
-    def get_iterator(self, split, batch_size, tgt_len, device, ext_len, mem_len=None):
+    def get_iterator(self, split, batch_size, tgt_len, device, ext_len, mem_len=None, diffp=False):
         if split == 'train':
             if self.dataset in ['ptb', 'wt2', 'wt103', 'enwik8', 'text8'] or self.dataset.startswith('olx_'):
-                data_iter = LMOrderedIterator(self.train, batch_size, tgt_len,
-                                              device=device, ext_len=ext_len, mem_len=mem_len)
+                if not diffp:
```
Comment (sytelus) [inline archai/nlp/datasets/corpus.py:206]: Please add comment why we need this separate paths.

File: archai/nlp/datasets/lm_dataset.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+from torch.utils.data import Dataset
+
+class LMDataset(Dataset):
```
Comment (sytelus) [inline archai/nlp/datasets/lm_dataset.py:3]: Please add comment why we need this new class and what it does

File: archai/nlp/models/hf_gpt2/hf_gpt2_utils/gpt2_lm_head_model_flex.py [hunk 3]
```diff
@@ -793,7 +793,8 @@ def forward(
             past_length = past_key_values[0][0].size(-2)
         if position_ids is None:
             position_ids = torch.arange(past_length, input_shape[-1] + past_length, dtype=torch.long, device=device)
-            position_ids = position_ids.unsqueeze(0).view(-1, input_shape[-1])
+            # Repeat position_ids so that Opacus interprets correctly the size of the batch
```
Comment (sytelus) [inline archai/nlp/models/hf_gpt2/hf_gpt2_utils/gpt2_lm_head_model_flex.py:796]: Is there any reference you can add in comment on why we need to do this? Also, would this effect non-diffp training?

File: archai/nlp/models/hf_gpt2/model_hf_gpt2.py [hunk 4]
```diff
@@ -32,6 +32,9 @@ def __init__(self, **kwargs) -> None:
 
         if self.config.tie_weight:
             self.model.tie_weights()
+            # Set the shared attribute so that Opacus knows that
+            # hte gradients must be summed and are not a new batch
+            self.model.lm_head.shared = True
```
Comment (sytelus) [inline archai/nlp/models/hf_gpt2/model_hf_gpt2.py:37]: Is this a special attribute Opacus specifically looks for? Also, could you please add comment with reference to more details?

File: archai/nlp/models/model_mixed_qat.py [hunk 5]
```diff
@@ -19,7 +19,7 @@ class MixedQATModel(ArchaiModel):
 
     """
 
-    def __init__(self, model: ArchaiModel, qat_weight: Optional[float] = 0.2) -> None:
+    def __init__(self, model: ArchaiModel, qat_weight: Optional[float] = 0.2, diffp: Optional[bool] = False) -> None:
```
Comment (sytelus) [inline archai/nlp/models/model_mixed_qat.py:22]: Why is this Optional?

**Review by sytelus — CHANGES_REQUESTED** [review]

File: archai/nlp/models/hf_gpt2/hf_gpt2_utils/gpt2_lm_head_model_flex.py (hunk 3 above)
Comment (caiom) [reply archai/nlp/models/hf_gpt2/hf_gpt2_utils/gpt2_lm_head_model_flex.py:796]: Will add more details in the comment. It does not affect non-diffp training.

File: archai/nlp/models/hf_gpt2/model_hf_gpt2.py (hunk 4 above)
Comment (caiom) [reply archai/nlp/models/hf_gpt2/model_hf_gpt2.py:37]: No. It is a attribute that my patch (Opacus changes) will look for. Will add more details.

File: archai/nlp/models/model_mixed_qat.py (hunk 5 above)
Comment (caiom) [reply archai/nlp/models/model_mixed_qat.py:22]: It is not, I was using Optional incorrectly. Will fix.

**Review by gugarosa — COMMENTED** [review]
Fixing conflict due to PR #77

**Comment by gugarosa** [comment]
Closing for now. Will do some refactoring in main classes and open this PR back with the modifications.

**Pull request closed without merging**